### ISSUE TITLE
feature request: Allow execute task with not default param inside script section

### ISSUE THREAD

**Issue by zestagio** [mumoshu/variant#73]
Hello,

Your tool is great but I faced with some problem. I don't know may be it already supported but I didn't find examples how to do it. Let's me explain. You have next example of usage in Readme file under **Getting Started** section:

```yaml
tasks:
  bar:
    script: |
      echo "dude"
  foo:
    parameters:
    - name: bar
      type: string
      description: "the bar"
    - name: environment
      type: string
      default: "heaven"
    script: |
      echo "Hello {{ get "bar" }} you are in the {{ get "environment" }}"
```
It works fine. But what about case if I want have **bar** task as parameterized and I need call it with not default param. For example:

```yaml
tasks:
  bar:
    parameters:
    - name: baz
      type: string
      default: "dude"
    script: |
      echo {{ get "baz" }}
  foo:
    parameters:
    - name: bar
      type: string
      description: "the bar"
    - name: environment
      type: string
      default: "heaven"
    script: |
      echo "Hello {{ get "bar" }} you are in the {{ get "environment" }}"
```

So, if I will call foo task `$ ./myfirstcmd foo` the **bar** task always works with default param and returns "dude", but I want send some argument when I call **bar** task inside script section of **foo** task, for example something like that:

```yaml
script: |
  echo "Hello {{ get "bar" set baz="folk" }} you are in the {{ get "environment" }}"
```

So, in general my idea is using result of parameterized task in another tasks. Is it possible?

**Comment by mumoshu** [comment]
@zestagio Hey! Thanks for trying variant ☺️ 

I think the best design depends on your actual script, but if it's really just that you want to somehow pass arbitrary parameter to `bar`'s `baz`, you can run your command like:

```
$ ./myfirstcmd foo --bar-baz folk
```

This isn't exactly what you've asked, but does this work for your use-case?

**Comment by zestagio** [comment]
@mumoshu

No, it's not very usable for me because I have one general task with parameter and the task returns result based on value of the parameter. Other tasks should call this general task inside script and use received result. According to your proposition I need call each task with `--bar-baz` parameter. It looks like overkill

**Comment by aroq** [comment]
@mumoshu, in fact, he needs to pass the output of parametrized task ("duplicate" in my example) as an input parameter of other tasks ("foo" & "bar"), and input parameter for "duplicate" task could be different during one variant call, so not possible to set it from command line.

Let me show it in the example:

```
tasks:
  duplicate:
    parameters:
    - name: input
      type: string
      default: ""
    script: |
      echo "{{.input}}_{{.input}}""
  foo:
    parameters:
    - name: argument
      type: string
    - name: duplicate
    script: |
      echo "Double {{.argument}} will be: {{.duplicate}}"
  bar:
    parameters:
    - name: argument
      type: string
    - name: duplicate
    steps:
    - script: echo "Other message: {{.duplicate}}"
    - script: echo "Other action"
  pipeline:
    parameters:
    - name: argument
      type: string
    - name: duplicate
    steps:    
    - task: foo
      arguments:
        argument: test
    - task: bar
      arguments:
        argument: test2   
```

I want to call it as:

```
variant pipeline

output:

Double test will be: test_test"
Double test2 will be: test2_test2"
Other action
```

Or probably we have some other ways to capture the output of parametrized tasks (with the parameters set directly during task call, and not with config or CMD flag) and re-use this output in other tasks?

**Comment by mumoshu** [comment]
@zestagio @aroq Hey! Thanks for your responses.

After consulting my mental model, I think @zestagio's `bar` and @aroq's `duplicate` in the `parameters` section aren't actually parameters, but rather "results" generated by calling another tasks(=`duplicate` with its own args).

Regarding the syntax, I would enhance each `step` to accept `name`, so that the result of the `step` is stored in the named var:

```yaml
tasks:
  foo:
    parameters:
    - name: argument
      type: string
    steps:
    - name: res
      task: duplicate
      arguments:
        input: "myinput"
    - script: |
      echo "Double {{.argument}} will be: {{.res}}"
```

We already have `steps`, each step taking `task + arguments` or `script` or various things. Adding `name` would allow us to refer the result of the previous step by name.

WDYT?

**Comment by zestagio** [comment]
@mumoshu hey!

It looks like good solution and seems it will solve my issue

**Comment by aroq** [comment]
@mumoshu this is a good solution which would cover our needs. In fact, it makes total sense, because we'll be able to pass the parameter "argument" defined in the task "foo" (from your example above) into "duplicate task", like:

```yaml
tasks:
  foo:
    parameters:
    - name: argument
      type: string
  steps:
  - name: res
    task: duplicate
    arguments:
      input: "{{.argument}}"
  - script: |
      echo "Double {{.argument}} will be: {{.res}}"
```

**Comment by mumoshu** [comment]
This has been released as part of v0.28.0. I'd appreciate it if you could test it and report back any issue you encounter. Thanks!
